File: vscode-extension/src/ide/settings/backendSettingsService.ts
/**
 * Backend Settings Service
 * 
 * Manages Claude and Codex backend configurations
 * Translated from jetbrains-plugin/src/main/kotlin/com/asakii/plugin/settings/BackendSettingsService.kt
 */

import * as vscode from 'vscode';
import { CodexSettings, codexSettings } from './codexSettings';

/**
 * Backend type enum
 */
export enum BackendType {
    CLAUDE = 'CLAUDE',
    CODEX = 'CODEX'
}

export const BackendTypeInfo: Record<BackendType, { displayName: string; icon: string }> = {
    [BackendType.CLAUDE]: { displayName: 'Claude', icon: '🤖' },
    [BackendType.CODEX]: { displayName: 'Codex', icon: '🔧' }
};

/**
 * Backend availability status
 */
export interface BackendAvailability {
    type: string;
    available: boolean;
    reason?: string;
}

/**
 * Backend config DTO (for frontend push)
 */
export interface BackendConfigDto {
    type: string;
    enabled: boolean;
    modelId?: string;
    modelProvider?: string;
    sandboxMode?: string;
    thinkingEnabled?: boolean;
    thinkingTokenBudget?: number;
    reasoningEffort?: string;
    reasoningSummary?: string;
}

/**
 * Backend settings state
 */
export interface BackendSettingsState {
    // Default backend type
    defaultBackend: string;
    
    // Claude config
    claudeEnabled: boolean;
    claudeModelId: string;
    claudeThinkingEnabled: boolean;
    claudeThinkingTokenBudget: number;
    claudeIncludePartialMessages: boolean;
    
    // Codex config
    codexEnabled: boolean;
    codexModelId: string;
    codexModelProvider: string;
    codexSandboxMode: string;
    codexReasoningEffort: string;
    codexReasoningSummary: string;
}

const DEFAULT_STATE: BackendSettingsState = {
    defaultBackend: BackendType.CLAUDE,
    claudeEnabled: true,
    claudeModelId: 'claude-sonnet-4-5-20250929',
    claudeThinkingEnabled: true,
    claudeThinkingTokenBudget: 8096,
    claudeIncludePartialMessages: true,
    codexEnabled: false,
    codexModelId: 'gpt-5.2-codex',
    codexModelProvider: 'openai',
    codexSandboxMode: 'workspace-write',
    codexReasoningEffort: 'xhigh',
    codexReasoningSummary: 'auto'
};

const CONFIG_KEY = 'claudeCodePlus.backend';

/**
 * Unified Backend Settings Service
 * 
 * Manages Claude and Codex backend configurations
 */
export class BackendSettingsService {
    private static instance: BackendSettingsService | undefined;
    private changeListeners: Set<(service: BackendSettingsService) => void> = new Set();

    private constructor() {}

    static getInstance(): BackendSettingsService {
        if (!this.instance) {
            this.instance = new BackendSettingsService();
        }
        return this.instance;
    }

    // ==================== State Management ====================

    /**
     * Get current state
     */
    getState(): BackendSettingsState {
        const config = vscode.workspace.getConfiguration(CONFIG_KEY);
        return {
            defaultBackend: config.get('defaultBackend', DEFAULT_STATE.defaultBackend),
            claudeEnabled: config.get('claudeEnabled', DEFAULT_STATE.claudeEnabled),
            claudeModelId: config.get('claudeModelId', DEFAULT_STATE.claudeModelId),
            claudeThinkingEnabled: config.get('claudeThinkingEnabled', DEFAULT_STATE.claudeThinkingEnabled),
            claudeThinkingTokenBudget: config.get('claudeThinkingTokenBudget', DEFAULT_STATE.claudeThinkingTokenBudget),
            claudeIncludePartialMessages: config.get('claudeIncludePartialMessages', DEFAULT_STATE.claudeIncludePartialMessages),
            codexEnabled: config.get('codexEnabled', DEFAULT_STATE.codexEnabled),
            codexModelId: config.get('codexModelId', DEFAULT_STATE.codexModelId),
            codexModelProvider: config.get('codexModelProvider', DEFAULT_STATE.codexModelProvider),
            codexSandboxMode: config.get('codexSandboxMode', DEFAULT_STATE.codexSandboxMode),
            codexReasoningEffort: config.get('codexReasoningEffort', DEFAULT_STATE.codexReasoningEffort),
            codexReasoningSummary: config.get('codexReasoningSummary', DEFAULT_STATE.codexReasoningSummary)
        };
    }

    // ==================== Listener Management ====================

    /**
     * Add change listener
     */
    addChangeListener(listener: (service: BackendSettingsService) => void): void {
        this.changeListeners.add(listener);
    }

    /**
     * Remove change listener
     */
    removeChangeListener(listener: (service: BackendSettingsService) => void): void {
        this.changeListeners.delete(listener);
    }

    /**
     * Notify all listeners
     */
    notifyChange(): void {
        this.changeListeners.forEach(listener => listener(this));
    }

    // ==================== Availability Detection ====================

    /**
     * Check if Claude backend is available
     */
    isClaudeAvailable(): boolean {
        return this.getState().claudeEnabled;
    }

    /**
     * Check if Codex backend is available
     */
    isCodexAvailable(): boolean {
        const state = this.getState();
        if (!state.codexEnabled) return false;
        return codexSettings.isValid();
    }

    /**
     * Get all backend availability status
     */
    getBackendAvailability(): BackendAvailability[] {
        const state = this.getState();
        const result: BackendAvailability[] = [];

        // Claude
        result.push({
            type: BackendType.CLAUDE.toLowerCase(),
            available: this.isClaudeAvailable(),
            reason: !state.claudeEnabled ? 'Claude is disabled' : undefined
        });

        // Codex
        let codexReason: string | undefined;
        if (!state.codexEnabled) {
            codexReason = 'Codex is disabled';
        } else if (!codexSettings.binaryPath) {
            codexReason = 'Codex binary path not configured';
        } else if (!codexSettings.isValid()) {
            codexReason = 'Codex binary not found or not executable';
        }

        result.push({
            type: BackendType.CODEX.toLowerCase(),
            available: this.isCodexAvailable(),
            reason: codexReason
        });

        return result;
    }

    /**
     * Get availability as JSON string
     */
    getBackendAvailabilityJson(): string {
        return JSON.stringify(this.getBackendAvailability(), null, 2);
    }

    // ==================== Config Getters ====================

    /**
     * Get default backend type
     */
    getDefaultBackend(): BackendType {
        const value = this.getState().defaultBackend;
        return BackendType[value as keyof typeof BackendType] ?? BackendType.CLAUDE;
    }

    /**
     * Set default backend type
     */
    async setDefaultBackend(type: BackendType): Promise<void> {
        await this.updateSetting('defaultBackend', type);
    }

    /**
     * Get Claude backend config DTO
     */
    getClaudeConfigDto(): BackendConfigDto {
        const state = this.getState();
        return {
            type: 'claude',
            enabled: state.claudeEnabled,
            modelId: state.claudeModelId,
            thinkingEnabled: state.claudeThinkingEnabled,
            thinkingTokenBudget: state.claudeThinkingTokenBudget
        };
    }

    /**
     * Get Codex backend config DTO
     */
    getCodexConfigDto(): BackendConfigDto {
        const state = this.getState();
        return {
            type: 'codex',
            enabled: state.codexEnabled,
            modelId: state.codexModelId,
            modelProvider: codexSettings.getState().modelProvider.toLowerCase(),
            sandboxMode: state.codexSandboxMode,
            reasoningEffort: state.codexReasoningEffort,
            reasoningSummary: state.codexReasoningSummary
        };
    }

    /**
     * Get all configs as JSON
     */
    getAllConfigsJson(): string {
        const configs = {
            defaultBackend: this.getState().defaultBackend.toLowerCase(),
            claude: this.getClaudeConfigDto(),
            codex: this.getCodexConfigDto(),
            availability: this.getBackendAvailability()
        };
        return JSON.stringify(configs, null, 2);
    }

    // ==================== Config Updates ====================

    /**
     * Update Claude config
     */
    async updateClaudeConfig(options: {
        enabled?: boolean;
        modelId?: string;
        thinkingEnabled?: boolean;
        thinkingTokenBudget?: number;
        includePartialMessages?: boolean;
    }): Promise<void> {
        const config = vscode.workspace.getConfiguration(CONFIG_KEY);
        
        if (options.enabled !== undefined) {
            await config.update('claudeEnabled', options.enabled, vscode.ConfigurationTarget.Workspace);
        }
        if (options.modelId !== undefined) {
            await config.update('claudeModelId', options.modelId, vscode.ConfigurationTarget.Workspace);
        }
        if (options.thinkingEnabled !== undefined) {
            await config.update('claudeThinkingEnabled', options.thinkingEnabled, vscode.ConfigurationTarget.Workspace);
        }
        if (options.thinkingTokenBudget !== undefined) {
            await config.update('claudeThinkingTokenBudget', options.thinkingTokenBudget, vscode.ConfigurationTarget.Workspace);
        }
        if (options.includePartialMessages !== undefined) {
            await config.update('claudeIncludePartialMessages', options.includePartialMessages, vscode.ConfigurationTarget.Workspace);
        }
        
        this.notifyChange();
    }

    /**
     * Update Codex config
     */
    async updateCodexConfig(options: {
        enabled?: boolean;
        modelId?: string;
        modelProvider?: string;
        sandboxMode?: string;
        reasoningEffort?: string;
        reasoningSummary?: string;
    }): Promise<void> {
        const config = vscode.workspace.getConfiguration(CONFIG_KEY);
        
        if (options.enabled !== undefined) {
            await config.update('codexEnabled', options.enabled, vscode.ConfigurationTarget.Workspace);
        }
        if (options.modelId !== undefined) {
            await config.update('codexModelId', options.modelId, vscode.ConfigurationTarget.Workspace);
        }
        if (options.modelProvider !== undefined) {
            await config.update('codexModelProvider', options.modelProvider, vscode.ConfigurationTarget.Workspace);
        }
        if (options.sandboxMode !== undefined) {
            await config.update('codexSandboxMode', options.sandboxMode, vscode.ConfigurationTarget.Workspace);
        }
        if (options.reasoningEffort !== undefined) {
            await config.update('codexReasoningEffort', options.reasoningEffort, vscode.ConfigurationTarget.Workspace);
        }
        if (options.reasoningSummary !== undefined) {
            await config.update('codexReasoningSummary', options.reasoningSummary, vscode.ConfigurationTarget.Workspace);
        }
        
        this.notifyChange();
    }

    // ==================== Convenience Properties ====================

    get defaultBackend(): string {
        return this.getState().defaultBackend;
    }

    get claudeEnabled(): boolean {
        return this.getState().claudeEnabled;
    }

    async setClaudeEnabled(value: boolean): Promise<void> {
        await this.updateClaudeConfig({ enabled: value });
    }

    get codexEnabled(): boolean {
        return this.getState().codexEnabled;
    }

    async setCodexEnabled(value: boolean): Promise<void> {
        await this.updateCodexConfig({ enabled: value });
    }

    // ==================== Private Helpers ====================

    private async updateSetting<K extends keyof BackendSettingsState>(
        key: K,
        value: BackendSettingsState[K]
    ): Promise<void> {
        const config = vscode.workspace.getConfiguration(CONFIG_KEY);
        await config.update(key, value, vscode.ConfigurationTarget.Workspace);
        this.notifyChange();
    }
}

// Export singleton instance
export const backendSettingsService = BackendSettingsService.getInstance();

File: jetbrains-plugin/src/main/kotlin/com/asakii/plugin/settings/BackendSettingsService.kt
package com.asakii.plugin.settings

import com.intellij.openapi.components.*
import com.intellij.openapi.project.Project
import com.intellij.util.xmlb.XmlSerializerUtil
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

/**
 * 后端类型枚举
 */
enum class BackendType(val displayName: String, val icon: String) {
    CLAUDE("Claude", "🤖"),
    CODEX("Codex", "🔧");

    companion object {
        fun fromName(name: String?): BackendType? {
            return entries.find { it.name.equals(name, ignoreCase = true) }
        }
    }
}

/**
 * 后端可用性状态
 */
@Serializable
data class BackendAvailability(
    val type: String,
    val available: Boolean,
    val reason: String? = null
)

/**
 * 后端配置 DTO（用于推送到前端）
 */
@Serializable
data class BackendConfigDto(
    val type: String,
    val enabled: Boolean,
    val modelId: String?,
    val modelProvider: String? = null,
    val sandboxMode: String? = null,
    val thinkingEnabled: Boolean = false,
    val thinkingTokenBudget: Int? = null,
    val reasoningEffort: String? = null,
    val reasoningSummary: String? = null
)

/**
 * 统一后端设置服务
 *
 * 管理 Claude 和 Codex 两个后端的配置，提供：
 * - 后端可用性检测
 * - 配置推送到前端
 * - 设置变更事件监听
 */
@State(
    name = "BackendSettings",
    storages = [Storage("backend-settings.xml")]
)
@Service(Service.Level.PROJECT)
class BackendSettingsService(private val project: Project) : PersistentStateComponent<BackendSettingsService.State> {

    data class State(
        // 默认后端类型
        var defaultBackend: String = BackendType.CLAUDE.name,

        // Claude 配置
        var claudeEnabled: Boolean = true,
        var claudeModelId: String = "claude-sonnet-4-6",
        var claudeThinkingEnabled: Boolean = true,
        var claudeThinkingTokenBudget: Int = 8096,
        var claudeIncludePartialMessages: Boolean = true,

        // Codex 配置
        var codexEnabled: Boolean = false,
        var codexModelId: String = "gpt-5.2-codex",
        var codexModelProvider: String = "openai",
        var codexSandboxMode: String = "workspace-write",
        var codexReasoningEffort: String = "xhigh",
        var codexReasoningSummary: String = "auto"
    )

    private var state = State()
    private val changeListeners = mutableListOf<(BackendSettingsService) -> Unit>()
    private val json = Json { ignoreUnknownKeys = true; prettyPrint = true }

    override fun getState(): State = state

    override fun loadState(state: State) {
        XmlSerializerUtil.copyBean(state, this.state)
    }

    // ==================== 监听器管理 ====================

    /**
     * 添加设置变更监听器
     */
    fun addChangeListener(listener: (BackendSettingsService) -> Unit) {
        changeListeners.add(listener)
    }

    /**
     * 移除设置变更监听器
     */
    fun removeChangeListener(listener: (BackendSettingsService) -> Unit) {
        changeListeners.remove(listener)
    }

    /**
     * 通知所有监听器设置已变更
     */
    fun notifyChange() {
        changeListeners.forEach { it(this) }
    }

    // ==================== 后端可用性检测 ====================

    /**
     * 检查 Claude 后端是否可用
     *
     * Claude 后端始终可用（使用 SDK）
     */
    fun isClaudeAvailable(): Boolean {
        return state.claudeEnabled
    }

    /**
     * 检查 Codex 后端是否可用
     *
     * 检查 Codex 二进制文件是否存在且可执行
     */
    fun isCodexAvailable(): Boolean {
        if (!state.codexEnabled) return false

        val codexSettings = CodexSettings.getInstance(project)
        return codexSettings.isValid()
    }

    /**
     * 获取所有后端的可用性状态
     */
    fun getBackendAvailability(): List<BackendAvailability> {
        val result = mutableListOf<BackendAvailability>()

        // Claude
        result.add(BackendAvailability(
            type = BackendType.CLAUDE.name.lowercase(),
            available = isClaudeAvailable(),
            reason = if (!state.claudeEnabled) "Claude is disabled" else null
        ))

        // Codex
        val codexSettings = CodexSettings.getInstance(project)
        val codexReason = when {
            !state.codexEnabled -> "Codex is disabled"
            codexSettings.binaryPath.isEmpty() -> "Codex binary path not configured"
            !codexSettings.isValid() -> "Codex binary not found or not executable"
            else -> null
        }
        result.add(BackendAvailability(
            type = BackendType.CODEX.name.lowercase(),
            available = isCodexAvailable(),
            reason = codexReason
        ))

        return result
    }

    /**
     * 获取可用性 JSON（用于 HTTP API）
     */
    fun getBackendAvailabilityJson(): String {
        return json.encodeToString(getBackendAvailability())
    }

    // ==================== 配置获取 ====================

    /**
     * 获取默认后端类型
     */
    fun getDefaultBackend(): BackendType {
        return BackendType.fromName(state.defaultBackend) ?: BackendType.CLAUDE
    }

    /**
     * 设置默认后端类型
     */
    fun setDefaultBackend(type: BackendType) {
        state.defaultBackend = type.name
        notifyChange()
    }

    /**
     * 获取 Claude 后端配置 DTO
     */
    fun getClaudeConfigDto(): BackendConfigDto {
        return BackendConfigDto(
            type = "claude",
            enabled = state.claudeEnabled,
            modelId = state.claudeModelId,
            thinkingEnabled = state.claudeThinkingEnabled,
            thinkingTokenBudget = state.claudeThinkingTokenBudget
        )
    }

    /**
     * 获取 Codex 后端配置 DTO
     */
    fun getCodexConfigDto(): BackendConfigDto {
        val codexSettings = CodexSettings.getInstance(project)
        return BackendConfigDto(
            type = "codex",
            enabled = state.codexEnabled,
            modelId = state.codexModelId,
            modelProvider = codexSettings.modelProvider.lowercase(),
            sandboxMode = state.codexSandboxMode,
            reasoningEffort = state.codexReasoningEffort,
            reasoningSummary = state.codexReasoningSummary
        )
    }

    /**
     * 获取所有后端配置（用于推送到前端）
     */
    fun getAllConfigsJson(): String {
        val configs = mapOf(
            "defaultBackend" to state.defaultBackend.lowercase(),
            "claude" to getClaudeConfigDto(),
            "codex" to getCodexConfigDto(),
            "availability" to getBackendAvailability()
        )
        return json.encodeToString(configs)
    }

    // ==================== 配置更新 ====================

    /**
     * 更新 Claude 配置
     */
    fun updateClaudeConfig(
        enabled: Boolean? = null,
        modelId: String? = null,
        thinkingEnabled: Boolean? = null,
        thinkingTokenBudget: Int? = null,
        includePartialMessages: Boolean? = null
    ) {
        enabled?.let { state.claudeEnabled = it }
        modelId?.let { state.claudeModelId = it }
        thinkingEnabled?.let { state.claudeThinkingEnabled = it }
        thinkingTokenBudget?.let { state.claudeThinkingTokenBudget = it }
        includePartialMessages?.let { state.claudeIncludePartialMessages = it }
        notifyChange()
    }

    /**
     * 更新 Codex 配置
     */
    fun updateCodexConfig(
        enabled: Boolean? = null,
        modelId: String? = null,
        modelProvider: String? = null,
        sandboxMode: String? = null,
        reasoningEffort: String? = null,
        reasoningSummary: String? = null
    ) {
        enabled?.let { state.codexEnabled = it }
        modelId?.let { state.codexModelId = it }
        modelProvider?.let { state.codexModelProvider = it }
        sandboxMode?.let { state.codexSandboxMode = it }
        reasoningEffort?.let { state.codexReasoningEffort = it }
        reasoningSummary?.let { state.codexReasoningSummary = it }
        notifyChange()
    }

    // ==================== 便捷属性 ====================

    var defaultBackend: String
        get() = state.defaultBackend
        set(value) {
            state.defaultBackend = value
            notifyChange()
        }

    var claudeEnabled: Boolean
        get() = state.claudeEnabled
        set(value) {
            state.claudeEnabled = value
            notifyChange()
        }

    var codexEnabled: Boolean
        get() = state.codexEnabled
        set(value) {
            state.codexEnabled = value
            notifyChange()
        }

    companion object {
        fun getInstance(project: Project): BackendSettingsService {
            return project.service<BackendSettingsService>()
        }
    }
}
